Fixes issue1248: Sort the tx inputs and outputs

hewison-chris opened the pull request:
The idea is to have a defined order for the inputs and outputs within a transaction. This is to make it deterministic so matches will more easily be made during nomination consensus and also make it not possible to re-arrange the transaction contents to create new hashes..

hewison-chris committed:
Fixes issue1248: Sort the tx inputs and outputs

The idea is to have a defined order for the inputs and outputs within a
transaction. This is to make it deterministic so matches will more
easily be made during nomination consensus.
It will also prevent attackers from re-arranging the inputs and outputs
to create different hashes.

hewison-chris committed:
Add ctors to `Transaction` struct

In preparation for issue 1248 where we will have sorted inputs and
outputs within a Transaction. Here we add `constructor`s for creating
`Transaction`s. When ctors are present then initializers are not allowed
so we replace them.

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -48,6 +49,61 @@ public enum TxType : ubyte
 {
+    @safe:
+
+    /// ctor with only outputs
+    this (in TxType type, in Output[] outputs) nothrow
+    {
+        this.type = type;
+        this.outputs = outputs.dup; // .sort.array; (FIXME: fix compilation error)
+    }
```
Accept a mutable value (let the caller `dup` if needed) and just call `sort` ?

hewison-chris replied:
Updated

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -48,6 +49,61 @@ public enum TxType : ubyte
+        this.payload = payload;
+    }
+
+    /// ctor with all params for immutable
+    this (immutable TxType type, immutable Input[] inputs, immutable Output[] outputs,
+        immutable DataPayload payload, immutable Height lock_height = Height(0)) immutable nothrow
```
```suggestion
    /// ctor with all params for immutable
    public this (in TxType type, inout Input[] inputs, inout Output[] outputs,
        inout DataPayload payload, in Height lock_height = Height(0)) inout nothrow
```

Something like this will also support `const`.

hewison-chris replied:
Updated

linked0 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -249,6 +306,12 @@ public struct Input
     }
+
+    /// Support for sorting
+    public int opCmp ( const typeof(this) rhs ) const nothrow @safe @nogc
```
```suggestion
    public int opCmp ( const typeof(this) rhs ) const @safe nothrow @nogc
```
I know "make @safe the first attribute" is our policy.

hewison-chris replied:
Thanks I updated

linked0 commented on source/agora/utils/WellKnownKeys.d:
```diff
@@ -1170,23 +1170,25 @@ static immutable CommonsBudget = KeyPair(PublicKey(Point([49, 189, 206, 20, 131,
-static immutable NODE6 = KeyPair(PublicKey(Point([217, 223, 183, 82, 128, 19, 204, 47, 227, 172, 199, 91, 151, 10, 85, 89, 206, 220, 236, 67, 60, 146, 140, 82, 239, 29, 150, 122, 71, 92, 159, 80])), SecretKey(Scalar([108, 48, 31, 186, 38, 25, 57, 33, 60, 210, 196, 151, 102, 61, 167, 8, 100, 128, 104, 220, 204, 27, 39, 67, 101, 67, 43, 119, 30, 25, 30, 4])));
-/// vald7: boa1xzvald7hxvgnzk50sy04ha7ezgyytxt5sgw323zy8dlj3ya2q40e6elltwq
-static immutable NODE7 = KeyPair(PublicKey(Point([153, 223, 183, 215, 51, 17, 49, 90, 143, 129, 31, 91, 247, 217, 18, 8, 69, 153, 116, 130, 29, 21, 68, 68, 59, 127, 40, 147, 170, 5, 95, 157])), SecretKey(Scalar([118, 108, 243, 157, 25, 205, 32, 124, 50, 230, 106, 83, 17, 206, 84, 190, 145, 205, 57, 39, 114, 49, 53, 82, 78, 242, 202, 178, 115, 229, 63, 8])));
+// val2: boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2 - SAQXRDHTWME4GUIVNYCKPN433VJ4BJP2L2T7UWHGSSW47VFC67EQFY3S
```
Good! I think "val2" looks better than "vald2". 
But this change should be another commit, I think.

hewison-chris replied:
Problem is it is needed for the sorted outputs in the freeze transaction of the GenesisBlock. If it is a separate commit to change the keys then all the unit tests have to be fixed twice.

linked0 replied:
Ah, I see. Ty.

hewison-chris replied:
In fact I am have some problem with it so will split it out into separate commit as you suggest.

linked0 reviewed (commented):
Good start.

hewison-chris committed:
Unit test comment correction for GenesisBlocks

The comments for testing `Test` and `Coinnet` `GenesisBlock`s were
swapped.

hewison-chris committed:
Remove commented out code

hewison-chris committed:
Network unit tests: Do not use random address

If we use random address then the transaction hash is non deterministic.
This means that the order of the transactions during the test run is
unknown when the test is written. This means that the test will
sometimes fail when the assert has assumed the transaction order.

hewison-chris commented:
Rebased and should be green.

MichaelKim20 commented:
@hewison-chris Is the rule that the first output is a refund kept?
This rule has been added in the PR below. I'd like to check, please.
https://github.com/bosagora/agora/pull/1664

MichaelKim20 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -183,6 +236,14 @@ public struct Output
     }
+
+    /// Support for sorting
+    public int opCmp ( const typeof(this) rhs ) const nothrow @safe @nogc
+    {
+        if (this.lock != rhs.lock)
+            return this.lock < rhs.lock ? -1 : 1;
+        return this.value.opCmp(rhs.value);
+    }
```
Is the rule that the first output is a refund kept?
This rule has been added in the PR below. I'd like to check, please.
https://github.com/bosagora/agora/pull/1664
When freezing 40,000 BOA from a UTXO containing 50,000 BOA, two outputs are generated. However, 40,000 BOA is not frozen, and other outputs may be frozen.

hewison-chris replied:
No, the rule is now that a single output may be less than the minimum freeze amount as long as there are more than one outputs.

hewison-chris replied:
Are you ok now with the change?

hewison-chris commented on source/agora/consensus/state/UTXOCache.d:
```diff
@@ -86,11 +87,14 @@ abstract class UTXOCache
-            // the first one is allowed to be under the minimum freezing amount.
+            // one is allowed to be under the minimum freezing amount.
             // If it is, it will not be considered frozen.
-            else if (tx.type == TxType.Freeze && tx.outputs.length >= 2 &&
-                     (idx == 0 && output.value < Amount.MinFreezeAmount))
-                this.add(utxo_hash, UTXO(height + 1, TxType.Payment, output));
+            else if (!frozen_refund_added && tx.type == TxType.Freeze
+                && tx.outputs.length >= 2 && output.value < Amount.MinFreezeAmount)
+                {
+                    this.add(utxo_hash, UTXO(height + 1, TxType.Payment, output));
+                    frozen_refund_added = true;
+                }
```
@MichaelKim20 : Here is the handling of it in our UTXOCache.

hewison-chris replied:
Including more lines

hewison-chris replied:
@MichaelKim20 : Here is handling in UTXOCache

hewison-chris replied:
Try again to highlight the code

MichaelKim20 reviewed (commented):
Yes, I checked. But please check if what I have in mind is right.
Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
About 50,000 BOA, which is the refund amount, is also frozen.

hewison-chris commented:
> Yes, I checked. But please check if what I have in mind is right.
> Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> About 50,000 BOA, which is the refund amount, is also frozen.

Only less than 40,000 BOA will be considered not frozen.
If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.

MichaelKim20 commented:
> > Yes, I checked. But please check if what I have in mind is right.
> > Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> > About 50,000 BOA, which is the refund amount, is also frozen.
> 
> Only less than 40,000 BOA will be considered not frozen.
> If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.

In order not to do so, we proceeded with the PR below.
https://github.com/bosagora/agora/pull/1664
Also, I think there is a problem with the fact that it is not clear which output is frozen.

hewison-chris commented:
> > > Yes, I checked. But please check if what I have in mind is right.
> > > Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> > > About 50,000 BOA, which is the refund amount, is also frozen.
> > 
> > 
> > Only less than 40,000 BOA will be considered not frozen.
> > If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.
> 
> In order not to do so, we proceeded with the PR below.
> #1664
> Also, I think there is a problem with the fact that it is not clear which output is frozen.

It is clear which output is unfrozen. The outputs are sorted and then the first output that is less than the minimum freeze amount.

MichaelKim20 commented:
@hewison-chris I'll take a little more time to organize my thoughts.
And I'll let you know my opinion again.

hewison-chris commented:
> @hewison-chris I'll take a little more time to organize my thoughts.
> And I'll let you know my opinion again.

We can discuss more next week if you like.

MichaelKim20 commented:
> > Yes, I checked. But please check if what I have in mind is right.
> > Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> > About 50,000 BOA, which is the refund amount, is also frozen.
> 
> Only less than 40,000 BOA will be considered not frozen.
> If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.

This method seems to be the most efficient solution in Agora.
There has been this discussion before, and I have asked the following questions.
If a user needs to split the funds first to freeze, he or she must wait (10 minutes) for the new block to create.
It can be difficult for users to understand and make it difficult to use a wallet.
In addition, unnecessary keypairs must be created in this process.
Therefore, I would like to ask you not only to consider the efficiency of Agora, but also to consider its convenience when used in Wallet and other tools.

hewison-chris commented:
> > > Yes, I checked. But please check if what I have in mind is right.
> > > Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> > > About 50,000 BOA, which is the refund amount, is also frozen.
> > 
> > 
> > Only less than 40,000 BOA will be considered not frozen.
> > If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.
> 
> This method seems to be the most efficient solution in Agora.
> There has been this discussion before, and I have asked the following questions.
> If a user needs to split the funds first to freeze, he or she must wait (10 minutes) for the new block to create.
> It can be difficult for users to understand and make it difficult to use a wallet.
> In addition, unnecessary keypairs must be created in this process.
> Therefore, I would like to ask you not only to consider the efficiency of Agora, but also to consider its convenience when used in Wallet and other tools.

@Geod24: Could we add a `refund` _field_ to the `Output` struct. Then there would be no need to limit the size of the unfrozen amount when the transaction type is `freeze`?

MichaelKim20 commented:
@Geod24 If we don't know the refund output, how can we change the fee using the replace function?
https://github.com/bosagora/agora/pull/1881

hewison-chris commented:
> > > > Yes, I checked. But please check if what I have in mind is right.
> > > > Only 40,000 BOA cannot be frozen on UTXO with 90,000 BOA.
> > > > About 50,000 BOA, which is the refund amount, is also frozen.
> > > 
> > > 
> > > Only less than 40,000 BOA will be considered not frozen.
> > > If you have UTXO with say 90,000 BOA then it needs to be split into two UTXO first if you want refund unfrozen.
> > 
> > 
> > This method seems to be the most efficient solution in Agora.
> > There has been this discussion before, and I have asked the following questions.
> > If a user needs to split the funds first to freeze, he or she must wait (10 minutes) for the new block to create.
> > It can be difficult for users to understand and make it difficult to use a wallet.
> > In addition, unnecessary keypairs must be created in this process.
> > Therefore, I would like to ask you not only to consider the efficiency of Agora, but also to consider its convenience when used in Wallet and other tools.
> 
> @Geod24: Could we add a `refund` _field_ to the `Output` struct. Then there would be no need to limit the size of the unfrozen amount when the transaction type is `freeze`?

The last commit of https://github.com/bosagora/agora/pull/2107 shows how it could work. (A test to fix still)

Geod24 commented on source/agora/consensus/data/Transaction.d:
```diff
@@ -249,6 +296,12 @@ public struct Input
     }
+
+    /// Support for sorting
+    public int opCmp ( const typeof(this) rhs ) const nothrow @safe @nogc
```
public int opCmp (in typeof(this) rhs) const nothrow @safe @nogc

Geod24 commented on source/agora/consensus/state/UTXOCache.d:
```diff
@@ -86,11 +87,14 @@ abstract class UTXOCache
                 this.add(utxo_hash, UTXO(height + 2016, tx.type, output));
             // Additionally, if a transaction has two or more outputs,
-            // the first one is allowed to be under the minimum freezing amount.
+            // one is allowed to be under the minimum freezing amount.
```
```suggestion
// two is allowed to be under the minimum freezing amount.
```
